File: rust_core/crates/strategy/tests/mrz_parity.py
from __future__ import annotations

import importlib
import json
import sys
from pathlib import Path
from typing import Any

import numpy as np

REPO_ROOT = Path(__file__).resolve().parents[4]
sys.path.insert(0, str(REPO_ROOT))
sys.path.insert(0, str(REPO_ROOT / "src" / "old"))

from src.old.backtest_engine.core.indicator_cache import IndicatorCache  # noqa: E402


class Candle:
    def __init__(self, close: float) -> None:
        self.open = close - 0.01
        self.high = close + 0.02
        self.low = close - 0.02
        self.close = close
        self.volume = 1.0


def make_candle(close: float) -> Candle:
    return Candle(close)


class Slice:
    def __init__(
        self,
        indicators: IndicatorCache,
        bid_candles: list[Candle],
        ask_candles: list[Candle],
        idx: int,
    ) -> None:
        self.indicators = indicators
        self._bid = bid_candles
        self._ask = ask_candles
        self.index = idx

    def latest(self, tf: str, price_type: str = "bid") -> Candle:
        if price_type == "bid":
            return self._bid[self.index]
        return self._ask[self.index]


def series_value(series, idx: int) -> float | None:
    if series is None:
        return None
    if idx < 0 or idx >= len(series):
        return None
    val = series.iloc[idx]
    if not np.isfinite(val):
        return None
    return float(val)


def signal_payload(signal: dict[str, Any] | None) -> dict[str, Any] | None:
    if not signal:
        return None
    return {
        "direction": str(signal.get("direction")),
        "entry": float(signal.get("entry")),
        "sl": float(signal.get("sl")),
        "tp": float(signal.get("tp")),
    }


def build_data(
    closes: list[float],
    h1_closes: list[float],
    ask_spread: float = 0.0002,
) -> dict[str, Any]:
    bid_candles = [make_candle(c) for c in closes]
    ask_candles = [make_candle(c + ask_spread) for c in closes]

    h1_candles = [make_candle(c) for c in h1_closes]
    rep = max(1, len(closes) // max(1, len(h1_candles)))
    h1_bid = []
    for candle in h1_candles:
        h1_bid.extend([candle] * rep)
    h1_bid = h1_bid[: len(closes)]
    h1_ask = [make_candle(c.close + ask_spread) for c in h1_bid]

    return {
        "bid_closes": closes,
        "ask_closes": [c + ask_spread for c in closes],
        "h1_bid_closes": h1_closes,
        "h1_ask_closes": [c + ask_spread for c in h1_closes],
        "bid_candles": bid_candles,
        "ask_candles": ask_candles,
        "h1_bid": h1_bid,
        "h1_ask": h1_ask,
    }


def build_payload(
    strategy_cls: type,
    params: dict[str, Any],
    closes: list[float],
    h1_closes: list[float],
) -> dict[str, Any]:
    data = build_data(closes, h1_closes)
    multi = {
        "M5": {"bid": data["bid_candles"], "ask": data["ask_candles"]},
        "H1": {"bid": data["h1_bid"], "ask": data["h1_ask"]},
    }
    cache = IndicatorCache(multi)
    idx = len(data["bid_closes"]) - 1
    slice_ = Slice(cache, data["bid_candles"], data["ask_candles"], idx)

    strat = build_strategy(strategy_cls, params)

    signals: dict[str, Any] = {}
    for scenario_id in range(1, 7):
        long_fn = getattr(strat, f"_evaluate_long_{scenario_id}")
        short_fn = getattr(strat, f"_evaluate_short_{scenario_id}")
        signals[f"{scenario_id}_long"] = signal_payload(
            long_fn(
                slice_,
                data["bid_candles"][idx],
                data["ask_candles"][idx],
            )
        )
        signals[f"{scenario_id}_short"] = signal_payload(
            short_fn(
                slice_,
                data["bid_candles"][idx],
                data["ask_candles"][idx],
            )
        )

    indicators = {
        "atr": series_value(cache.atr("M5", "bid", params["atr_length"]), idx),
        "ema": series_value(cache.ema("M5", "bid", params["ema_length"]), idx),
        "kalman_z": series_value(
            cache.kalman_zscore(
                "M5",
                "bid",
                window=params["window_length"],
                R=params["kalman_r"],
                Q=params["kalman_q"],
            ),
            idx,
        ),
        "kalman_garch_z": series_value(
            cache.kalman_garch_zscore(
                "M5",
                "bid",
                R=params["kalman_r"],
                Q=params["kalman_q"],
                alpha=params["garch_alpha"],
                beta=params["garch_beta"],
                omega=params["garch_omega"],
                use_log_returns=params["garch_use_log_returns"],
                scale=params["garch_scale"],
                min_periods=params["garch_min_periods"],
                sigma_floor=params["garch_sigma_floor"],
            ),
            idx,
        ),
    }

    bb_upper, bb_mid, bb_lower = cache.bollinger(
        "M5", "bid", period=params["b_b_length"], std_factor=params["std_factor"]
    )
    indicators["bollinger"] = {
        "upper": series_value(bb_upper, idx),
        "middle": series_value(bb_mid, idx),
        "lower": series_value(bb_lower, idx),
    }

    h1_kz = cache.kalman_zscore_stepwise(
        "H1",
        "bid",
        window=params["scenario6_params"]["H1"]["window_length"],
        R=params["scenario6_params"]["H1"]["kalman_r"],
        Q=params["scenario6_params"]["H1"]["kalman_q"],
    )
    h1_upper, h1_mid, h1_lower = cache.bollinger_stepwise(
        "H1",
        "bid",
        period=params["scenario6_params"]["H1"]["b_b_length"],
        std_factor=params["scenario6_params"]["H1"]["std_factor"],
    )
    h1_price = cache.get_closes("H1", "bid")
    indicators["scenario6_debug"] = {
        "kalman_z_step": series_value(h1_kz, idx),
        "lower": series_value(h1_lower, idx),
        "price": series_value(h1_price, idx),
        "upper": series_value(h1_upper, idx),
        "middle": series_value(h1_mid, idx),
    }
    chain_ok, chain_meta = strat._scenario6_evaluate_chain(slice_, idx, "long")
    indicators["scenario6_chain_ok"] = chain_ok
    indicators["scenario6_chain_meta"] = chain_meta

    return {
        "index": idx,
        "params": params,
        "data": {
            "bid_closes": data["bid_closes"],
            "ask_closes": data["ask_closes"],
            "h1_bid_closes": data["h1_bid_closes"],
            "h1_ask_closes": data["h1_ask_closes"],
        },
        "signals": signals,
        "indicators": indicators,
    }


def build_strategy(cls, params: dict[str, Any]) -> Any:
    strat = cls.__new__(cls)
    strat.symbol = "EURUSD"
    strat.timeframe = "M5"
    strat.pip_size = 0.0001

    strat.atr_length = int(params["atr_length"])
    strat.atr_mult = float(params["atr_mult"])
    strat.b_b_length = int(params["b_b_length"])
    strat.std_factor = float(params["std_factor"])
    strat.window_length = int(params["window_length"])
    strat.z_score_long = float(params["z_score_long"])
    strat.z_score_short = float(params["z_score_short"])
    strat.ema_length = int(params["ema_length"])
    strat.kalman_r = float(params["kalman_r"])
    strat.kalman_q = float(params["kalman_q"])
    strat.tp_min_distance = float(params["tp_min_distance"])

    strat.direction_filter = "both"
    strat.enabled_scenarios = {1, 2, 3, 4, 5, 6}
    strat.allowed_labels = set()
    strat.use_position_manager = False
    strat.position_manager = None

    strat.htfA_tf = str(params.get("htf_tf", "NONE")).upper()
    strat.htfA_ema = int(params.get("htf_ema", 0))
    strat.htfA_filter = str(params.get("htf_filter", "none")).lower()
    strat.htfB_tf = str(params.get("extra_htf_tf", "NONE")).upper()
    strat.htfB_ema = int(params.get("extra_htf_ema", 0))
    strat.htfB_filter = str(params.get("extra_htf_filter", "none")).lower()

    strat.garch_alpha = float(params["garch_alpha"])
    strat.garch_beta = float(params["garch_beta"])
    strat.garch_omega = float(params["garch_omega"])
    strat.garch_use_log_returns = bool(params.get("garch_use_log_returns", True))
    strat.garch_scale = float(params.get("garch_scale", 100.0))
    strat.garch_min_periods = int(params.get("garch_min_periods", 50))
    strat.garch_sigma_floor = float(params.get("garch_sigma_floor", 1e-6))
    strat.local_z_lookback = int(params.get("local_z_lookback", 0))

    strat.intraday_vol_feature = str(params["intraday_vol_feature"]).lower()
    strat.intraday_vol_cluster_window = int(params["intraday_vol_cluster_window"])
    strat.intraday_vol_cluster_k = int(params["intraday_vol_cluster_k"])
    strat.intraday_vol_log_transform = bool(
        params.get("intraday_vol_log_transform", True)
    )
    strat.intraday_vol_min_points = int(params.get("intraday_vol_min_points", 1))
    strat.intraday_vol_garch_lookback = int(
        params.get("intraday_vol_garch_lookback", 500)
    )
    strat.cluster_hysteresis_bars = int(params.get("cluster_hysteresis_bars", 0))
    strat.intraday_vol_allowed = list(params.get("intraday_vol_allowed", []))
    strat._vol_cluster_cache = {}
    strat._local_z_cache = {}

    scenario6_mode = str(params.get("scenario6_mode", "all")).lower()
    strat.scenario6_mode = scenario6_mode if scenario6_mode in ("all", "any") else "all"
    strat.scenario6_timeframes = [
        str(tf).upper() for tf in params.get("scenario6_timeframes", [])
    ]
    strat.scenario6_params = params.get("scenario6_params", {})

    strat.news_filter = None
    strat.portfolio = None
    return strat


def main() -> None:
    mrz_module = importlib.import_module(
        "src.old.strategies.mean_reversion_z_score.backtest.backtest_strategy"
    )
    MeanReversionZScoreStrategy = mrz_module.MeanReversionZScoreStrategy
    params = {
        "atr_length": 3,
        "atr_mult": 2.0,
        "b_b_length": 3,
        "std_factor": 1.0,
        "window_length": 3,
        "z_score_long": -0.5,
        "z_score_short": 0.5,
        "ema_length": 3,
        "kalman_r": 0.5,
        "kalman_q": 0.1,
        "garch_alpha": 0.1,
        "garch_beta": 0.8,
        "garch_omega": 0.00001,
        "garch_min_periods": 2,
        "garch_use_log_returns": True,
        "garch_scale": 100.0,
        "garch_sigma_floor": 1e-6,
        "tp_min_distance": 0.0001,
        "htf_filter": "none",
        "htf_tf": "H1",
        "htf_ema": 3,
        "extra_htf_tf": "NONE",
        "extra_htf_filter": "none",
        "scenario6_mode": "all",
        "scenario6_timeframes": ["H1"],
        "scenario6_params": {
            "H1": {
                "window_length": 3,
                "b_b_length": 3,
                "std_factor": 0.5,
                "kalman_r": 0.5,
                "kalman_q": 0.1,
                "z_score_long": -0.5,
                "z_score_short": 0.5,
            }
        },
        "intraday_vol_feature": "atr_points",
        "intraday_vol_cluster_window": 5,
        "intraday_vol_cluster_k": 1,
        "intraday_vol_allowed": [
            "low",
            "mid",
            "high",
            "very_high",
            "extreme",
        ],
        "intraday_vol_min_points": 1,
        "cluster_hysteresis_bars": 0,
    }

    long_closes = [1.0] * 8 + [0.95, 0.94, 0.93, 0.92]
    long_h1 = [1.0, 0.96, 0.92]
    short_closes = [1.0] * 8 + [1.06, 1.07, 1.08, 1.09]
    short_h1 = [1.0, 1.04, 1.08]

    payload = {
        "cases": {
            "long": build_payload(
                MeanReversionZScoreStrategy, params, long_closes, long_h1
            ),
            "short": build_payload(
                MeanReversionZScoreStrategy, params, short_closes, short_h1
            ),
        }
    }

    json.dump(payload, sys.stdout)


if __name__ == "__main__":
    main()

File: rust_core/crates/strategy/tests/mrz_parity.rs
use std::collections::HashMap;
use std::path::PathBuf;
use std::process::Command;

use omega_indicators::{IndicatorCache, IndicatorParams, IndicatorSpec, MultiTfIndicatorCache};
use omega_strategy::{BarContext, MeanReversionZScore, MrzParams, Strategy};
use omega_types::{Candle, Direction, Timeframe};
use serde::Deserialize;

#[derive(Debug, Deserialize)]
struct ParityPayload {
    index: usize,
    params: serde_json::Value,
    data: ParityData,
    signals: HashMap<String, Option<SignalPayload>>,
    indicators: IndicatorSnapshot,
}

#[derive(Debug, Deserialize)]
struct ParitySuite {
    cases: HashMap<String, ParityPayload>,
}

#[derive(Debug, Deserialize)]
struct ParityData {
    bid_closes: Vec<f64>,
    ask_closes: Vec<f64>,
    h1_bid_closes: Vec<f64>,
    h1_ask_closes: Vec<f64>,
}

#[derive(Debug, Deserialize, Clone)]
struct SignalPayload {
    direction: String,
    entry: f64,
    sl: f64,
    tp: f64,
}

#[derive(Debug, Deserialize)]
struct IndicatorSnapshot {
    kalman_z: Option<f64>,
    kalman_garch_z: Option<f64>,
    bollinger: BollingerSnapshot,
    scenario6_debug: Scenario6Debug,
}

#[derive(Debug, Deserialize)]
struct Scenario6Debug {
    kalman_z_step: Option<f64>,
    lower: Option<f64>,
    price: Option<f64>,
    upper: Option<f64>,
    middle: Option<f64>,
}

#[derive(Debug, Deserialize)]
struct BollingerSnapshot {
    upper: Option<f64>,
    middle: Option<f64>,
    lower: Option<f64>,
}

#[test]
fn test_mrz_python_parity_scenarios() {
    let suite = run_python_parity();

    for (case_key, direction) in [("long", "long"), ("short", "short")] {
        let payload = suite
            .cases
            .get(case_key)
            .unwrap_or_else(|| panic!("missing {case_key} payload"));
        let base_params: MrzParams =
            serde_json::from_value(payload.params.clone()).expect("invalid MRZ params");

        let primary_bid = build_candles(&payload.data.bid_closes, 0, 1);
        let primary_ask = build_candles(&payload.data.ask_closes, 0, 1);

        let h1_bid = build_candles(&payload.data.h1_bid_closes, 0, 4);
        let h1_ask = build_candles(&payload.data.h1_ask_closes, 0, 4);

        let mut indicator_cache = IndicatorCache::new();
        insert_kalman_z(
            &mut indicator_cache,
            payload.index,
            base_params.window_length,
            base_params.kalman_r,
            base_params.kalman_q,
            payload.indicators.kalman_z,
        );
        insert_kalman_garch_z(
            &mut indicator_cache,
            payload.index,
            base_params.window_length,
            KalmanGarchSnapshot {
                r: base_params.kalman_r,
                q: base_params.kalman_q,
                alpha: base_params.garch_alpha,
                beta: base_params.garch_beta,
                omega: base_params.garch_omega,
                use_log_returns: base_params.garch_use_log_returns,
                scale: base_params.garch_scale,
                min_periods: base_params.garch_min_periods,
                sigma_floor: base_params.garch_sigma_floor,
                value: payload.indicators.kalman_garch_z,
            },
        );
        insert_bollinger(
            &mut indicator_cache,
            payload.index,
            base_params.b_b_length,
            base_params.std_factor,
            &payload.indicators.bollinger,
        );

        insert_stepwise_kalman_z(
            &mut indicator_cache,
            payload.index,
            "H1",
            base_params.window_length,
            base_params.kalman_r,
            base_params.kalman_q,
            payload.indicators.scenario6_debug.kalman_z_step,
        );
        insert_stepwise_bollinger(
            &mut indicator_cache,
            payload.index,
            "H1",
            base_params.b_b_length,
            base_params.std_factor,
            &payload.indicators.scenario6_debug,
        );
        insert_tf_close(
            &mut indicator_cache,
            payload.index,
            "H1",
            payload.indicators.scenario6_debug.price,
        );

        let multi_tf = std::cell::RefCell::new(MultiTfIndicatorCache::new(
            Timeframe::M5,
            primary_bid.clone(),
            primary_ask.clone(),
            vec![(Timeframe::H1, h1_bid, h1_ask)],
        ));

        for scenario_id in 1..=6 {
            let mut params = base_params.clone();
            params.enabled_scenarios = vec![scenario_id];
            let mut strategy = MeanReversionZScore::new(params);

            let bid = &primary_bid[payload.index];
            let ask = &primary_ask[payload.index];
            let ctx = BarContext::new(payload.index, bid.timestamp_ns, bid, ask, &indicator_cache)
                .with_multi_tf(&multi_tf);

            let actual = strategy.on_bar(&ctx);
            let expected_key = format!("{scenario_id}_{direction}");
            let expected = payload.signals.get(&expected_key).cloned().unwrap_or(None);
            assert_signal_matches(case_key, scenario_id, expected, actual);
        }
    }
}

fn build_candles(closes: &[f64], start_ts: i64, step: i64) -> Vec<Candle> {
    closes
        .iter()
        .enumerate()
        .map(|(idx, close)| Candle {
            timestamp_ns: start_ts + (idx as i64 * step),
            open: close - 0.01,
            high: close + 0.02,
            low: close - 0.02,
            close: *close,
            volume: 1.0,
        })
        .collect()
}

fn insert_kalman_z(
    cache: &mut IndicatorCache,
    idx: usize,
    window: usize,
    r: f64,
    q: f64,
    value: Option<f64>,
) {
    let spec = IndicatorSpec::new(
        "KALMAN_Z",
        IndicatorParams::Kalman {
            window,
            r_x1000: (r * 1000.0).round() as u32,
            q_x1000: (q * 1000.0).round() as u32,
        },
    );
    insert_scalar(cache, spec, idx, value);
}

struct KalmanGarchSnapshot {
    r: f64,
    q: f64,
    alpha: f64,
    beta: f64,
    omega: f64,
    use_log_returns: bool,
    scale: f64,
    min_periods: usize,
    sigma_floor: f64,
    value: Option<f64>,
}

fn insert_kalman_garch_z(
    cache: &mut IndicatorCache,
    idx: usize,
    window: usize,
    snapshot: KalmanGarchSnapshot,
) {
    let spec = IndicatorSpec::new(
        "KALMAN_GARCH_Z",
        IndicatorParams::KalmanGarch {
            window,
            r_x1000: (snapshot.r * 1000.0).round() as u32,
            q_x1000: (snapshot.q * 1000.0).round() as u32,
            alpha_x1000: (snapshot.alpha * 1000.0).round() as u32,
            beta_x1000: (snapshot.beta * 1000.0).round() as u32,
            omega_x1000000: (snapshot.omega * 1_000_000.0).round() as u32,
            use_log_returns: snapshot.use_log_returns,
            scale_x100: (snapshot.scale * 100.0).round() as u32,
            min_periods: snapshot.min_periods,
            sigma_floor_x1e8: (snapshot.sigma_floor * 1e8).round() as u32,
        },
    );
    insert_scalar(cache, spec, idx, snapshot.value);
}

fn insert_bollinger(
    cache: &mut IndicatorCache,
    idx: usize,
    period: usize,
    std_factor: f64,
    snapshot: &BollingerSnapshot,
) {
    let spec = IndicatorSpec::new(
        "BOLLINGER",
        IndicatorParams::Bollinger {
            period,
            std_factor_x100: (std_factor * 100.0).round() as u32,
        },
    );
    insert_scalar(cache, spec.with_output_suffix("upper"), idx, snapshot.upper);
    insert_scalar(
        cache,
        spec.with_output_suffix("middle"),
        idx,
        snapshot.middle,
    );
    insert_scalar(cache, spec.with_output_suffix("lower"), idx, snapshot.lower);
}

fn insert_stepwise_kalman_z(
    cache: &mut IndicatorCache,
    idx: usize,
    timeframe: &str,
    window: usize,
    r: f64,
    q: f64,
    value: Option<f64>,
) {
    let spec = IndicatorSpec::new(
        format!("KALMAN_Z_{timeframe}").as_str(),
        IndicatorParams::Kalman {
            window,
            r_x1000: (r * 1000.0).round() as u32,
            q_x1000: (q * 1000.0).round() as u32,
        },
    );
    insert_scalar(cache, spec, idx, value);
}

fn insert_stepwise_bollinger(
    cache: &mut IndicatorCache,
    idx: usize,
    timeframe: &str,
    period: usize,
    std_factor: f64,
    snapshot: &Scenario6Debug,
) {
    let spec = IndicatorSpec::new(
        format!("BOLLINGER_{timeframe}").as_str(),
        IndicatorParams::Bollinger {
            period,
            std_factor_x100: (std_factor * 100.0).round() as u32,
        },
    );
    insert_scalar(cache, spec.with_output_suffix("upper"), idx, snapshot.upper);
    insert_scalar(
        cache,
        spec.with_output_suffix("middle"),
        idx,
        snapshot.middle,
    );
    insert_scalar(cache, spec.with_output_suffix("lower"), idx, snapshot.lower);
}

fn insert_tf_close(cache: &mut IndicatorCache, idx: usize, timeframe: &str, value: Option<f64>) {
    let spec = IndicatorSpec::new(
        format!("CLOSE_{timeframe}").as_str(),
        IndicatorParams::Period(1),
    );
    insert_scalar(cache, spec, idx, value);
}

fn insert_scalar(cache: &mut IndicatorCache, spec: IndicatorSpec, idx: usize, value: Option<f64>) {
    let mut values = vec![f64::NAN; idx + 1];
    if let Some(val) = value {
        values[idx] = val;
    }
    cache.insert(spec, values);
}

fn assert_signal_matches(
    case_key: &str,
    scenario_id: u8,
    expected: Option<SignalPayload>,
    actual: Option<omega_types::Signal>,
) {
    let tolerance = 1e-5;
    match (expected, actual) {
        (None, None) => {}
        (Some(exp), Some(act)) => {
            let expected_dir = match exp.direction.as_str() {
                "long" => Direction::Long,
                "short" => Direction::Short,
                other => panic!("Scenario {scenario_id}: unknown direction {other}"),
            };
            assert_eq!(
                act.direction, expected_dir,
                "Case {case_key} scenario {scenario_id} direction"
            );
            assert_close(
                exp.entry,
                act.entry_price,
                tolerance,
                case_key,
                scenario_id,
                "entry",
            );
            assert_close(
                exp.sl,
                act.stop_loss,
                tolerance,
                case_key,
                scenario_id,
                "sl",
            );
            assert_close(
                exp.tp,
                act.take_profit,
                tolerance,
                case_key,
                scenario_id,
                "tp",
            );
        }
        (None, Some(act)) => panic!(
            "Case {case_key} scenario {scenario_id}: expected no signal, got {:?}",
            act
        ),
        (Some(exp), None) => panic!(
            "Case {case_key} scenario {scenario_id}: expected signal {:?}, got None",
            exp
        ),
    }
}

fn assert_close(
    expected: f64,
    actual: f64,
    tol: f64,
    case_key: &str,
    scenario_id: u8,
    label: &str,
) {
    let diff = (expected - actual).abs();
    assert!(
        diff <= tol,
        "Case {case_key} scenario {scenario_id} {label} diff {diff} exceeds {tol}: {expected} vs {actual}"
    );
}

fn run_python_parity() -> ParitySuite {
    let python = std::env::var("PYTHON").unwrap_or_else(|_| "python3".to_string());
    let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let script_path = manifest_dir.join("tests").join("mrz_parity.py");

    let output = Command::new(&python)
        .arg(script_path)
        .output()
        .expect("Failed to run mrz_parity.py");

    if !output.status.success() {
        panic!(
            "mrz_parity.py failed: {}",
            String::from_utf8_lossy(&output.stderr)
        );
    }

    serde_json::from_slice(&output.stdout).expect("Failed to parse MRZ parity payload")
}
